src/py/mains/run.py

```diff
@@ -111,31 +120,141 @@ def wait_for_file(filename, task, job_name = ""):
+    """(This is the old way.)
+    We could be in the thread or sub-process which spawned a qsub job,
+    so we must check for the shutdown_event.
+    """
+    wait_time = .2
+    while 1:
+        time.sleep(wait_time)
+        wait_time = 10 if wait_time > 10 else wait_time + .2
+        status = check_for_file_or_exit(filename, task, job_name)
+        if status != Waiter.StillRunning:
+            break
+
+def wait_for_file_to_exist(fn):
+    """This can wait forever.
+    """
+    wait_time = .2
+    while 1:
+        time.sleep(wait_time)
+        wait_time = 10 if wait_time > 10 else wait_time + .2
+        if os.path.exists(fn):
+            break
+
+def wait_for_file_with_progress(filename, progress_fn, task, job_name):
+    """Continue as long as we detect progress.
+    Ignore the .exit file. Instead, watch for changes to progress_fn.
+    We could be in the thread or sub-process which spawned a qsub job,
+    so we must check for the shutdown_event.
+    (I have so far made as few changes to the previous version as possible,
+     but someday this should be re-architected. I favor a separate, queryable
+     process.)
+    """
+    # First, wait for progress_fn to exist at all.
+    # We assume that any submitted job will eventually start. If not, the user
+    # must hit Ctrl-C, because we really cannot know how to long to wait.
+    fc_run_logger.debug('Waiting for {}'.format(progress_fn))
+    wait_for_file_to_exist(progress_fn)
+    fc_run_logger.debug('Found {}'.format(progress_fn))
+
+    # Next, decide how long to allow between updates.
+    max_update_delta = datetime.timedelta(seconds=MAX_UPDATE_DELTA_S)
+    last_progress = os.path.getmtime(progress_fn)
+    last_update = datetime.datetime.now()
+    wait_time = .2
+
+    # Loop until we exceed max_update_delta, or until we find something.
+    while 1:
+        time.sleep(wait_time)
+        wait_time = 10 if wait_time > 10 else wait_time + .2
+        status = check_for_file_or_exit(filename, task, job_name)
+        if status != Waiter.StillRunning:
             break
+        latest_progress = os.path.getmtime(progress_fn)
+        latest_update = datetime.datetime.now()
+        #fc_run_logger.info('last={}, latest={}, last={}, latest={}, diff={}, cond={}'.format(
+        #    last_progress, latest_progress,
+        #    last_update, latest_update,
+        #    latest_update - last_update,
+        #    latest_update - last_update > max_update_delta,
+        #))
+        if last_progress == latest_progress:
+            if latest_update - last_update > max_update_delta:
+                # Assume the job stopped, but cancel if not.
+                cancel_job(job_name, support.job_type)
+                break
+            else:
+                continue
+        last_progress = latest_progress
+        last_update = latest_update
+
+# Compare nap to MAX_UPDATE_DELTA_S.
+# nap should be less!
+WRAPPER_SCRIPT = """
+progress() {
```

Comment:
I also tried a version which creates a separate progress script to run in the background. That makes it slightly easier to avoid dangling processes. However, if someday runs `kill -9 PID`, there is nothing we can do anyway. Whoever is killing out submitted jobs (if that is what's happening) is responsible for running `kill -9 -PID`, with a negative PID to kill the entire process group.